File: sorting/radix.js
// NOTE: Seems to be broken. This counting sort implementation seems not to be stable
const objects = [{ date: '2-20-18', val: 4, }, { date: '4-20-17', val: 6, }, { date: '3-20-17', val: 5, }, { date: '6-2-17', val: 7, }, { date: '2-18-18', val: 3, }, { date: '1-20-17', val: 2, }, { date: '6-3-15', val: 1, }];
radixSort(objects);
console.log('Objects using Radix Sort:', objects);

function radixSort(arr) {
  for (let i = 0; i < arr.length; i++) {
    const [ month, day, year ] = arr[i].date.split('-');
    arr[i].month = Number(month);
    arr[i].day = Number(day);
    arr[i].year = Number(year);
  }

  countingSort(arr, 31, 'day');
  console.log('post day:', arr)
  countingSort(arr, 12, 'month');
  console.log('post month:', arr)
  countingSort(arr, 18, 'year');

  for (let i = 0; i < arr.length; i++) {
    delete arr[i].day;
    delete arr[i].month;
    delete arr[i].year;
  }
}

// We are guaranteed that each element of 'nums' is an integer x where 0 <= x <= k.
function countingSort(objects, k, field) {
  const out = Array(objects.length);
  const store = Array(k+1).fill(0);
  // make each idx of 'store' hold the # of elements in 'nums' that equal that idx
  for (const o of objects) {
    store[o[field]]++;
  }
  // make each idx of 'store' hold the # of elements that come before it
  for (let i = 1; i < store.length; i++) {
    store[i] += store[i-1];
  }
  // place elements in correct positions in 'out' array
  for (let i = 0; i < objects.length; i++) {
    const o = objects[i];
    out[store[o[field]]-1] = o;
    store[o[field]]--;
  }
  // overwrite input array using 'out' array, which is sorted
  for (let i = 0; i < out.length; i++) {
    objects[i] = out[i];
  }
}

console.log('\n=== counting sort test ===')
const arr1 = [{ val: 4, }, { val: 5, }, { val: 6, }, { val: 3, }, { val: 2, }, { val: 1, }];
countingSort(arr1, 7, 'val');
console.log('arr1:', arr1);

File: sorting/radix.go
package main

import (
	"fmt"
)

// COME BACK TO THIS
func RadixSort(nums []string, d int) {
	for i := d - 1; i >= 0; i-- {

	}
}

// CountingSort will sort a list of integers using the counting sort algorithm.
//
// Big O: O(k+n), where n is the size of the list containing integers in the range 0-k
func CountingSort(nums []int, k int) {
	if len(nums) <= 1 {
		return
	}

	out := make([]int, len(nums))
	store := make([]int, k+1)

	// each elem in store now contains the number of elements equal to v
	for _, v := range nums {
		store[v]++
	}
	// each elem in store now contains the number of elements <= v
	for i := 1; i < len(store); i++ {
		store[i] += store[i-1]
	}
	// place each element in nums in the correct place in the output array
	for _, v := range nums {
		out[store[v]-1] = v
		store[v]--
	}

	for i, v := range out {
		nums[i] = v
	}
}

func main() {
	toSort := []string{"178", "399", "112", "202", "434", "583", "723", "533", "201"}
	RadixSort(toSort, 3)
	fmt.Println(toSort)
}
